web_monitoring/diffing_server.py

```diff
@@ -67,17 +69,27 @@ def get(self, differ):
         try:
             func = self.differs[differ]
         except KeyError:
-            self.send_error(404)
+            self.send_error(404, reason = f'Unknown diffing method: `{differ}`. You can get a list of supported differs from the `/` endpoint.')
             return
 
         # If params repeat, take last one. Decode bytes into unicode strings.
         query_params = {k: v[-1].decode() for k, v in
                         self.request.arguments.items()}
-        a = query_params.pop('a')
-        b = query_params.pop('b')
-
+        try:
+            a = query_params.pop('a')
+            b = query_params.pop('b')
+        except KeyError:
+            self.send_error(
+                        400, reason = 'Malformed request. You must provide a URL as the value for both `a` and `b` query parameters.')
+            return
         # Fetch server response for URLs a and b.
-        res_a, res_b = yield [client.fetch(a), client.fetch(b)]
+        res_a, res_b = yield [client.fetch(a, raise_error=False), client.fetch(b, raise_error=False)]
+
+        try:
+            self.check_response_for_error(res_a)
+            self.check_response_for_error(res_b)
+        except tornado.httpclient.HTTPError:
+            return
```

Comment:
I’m pretty sure this is a logic error. In `check_response_for_error()`, we have:

```py
if response.error is not None:
    try:
        response.rethrow()
    except (ValueError, IOError):
        # send back the right error response
        raise tornado.httpclient.HTTPError(0)
    # But we didn’t handle tornado.httpclient.HTTPError errors!
```

If `response.rethrow()` raised a `tornado.httpclient.HTTPError`, it won’t ever get an error response because we didn’t handle it inside `check_response_for_error()`, but we still stopped here in `DiffHandler.get()` as if we did. It basically stops here, but never sends the error message. You can try this by adding a test:

```py
def test_a_is_404(self):
        response = self.fetch(f'/html_token?format=json&include=all&a=http://httpstat.us/404&b=https://example.org')
        print(f'STATUS:: {response.code}')
        print('BODY:::: "%s"' % response.body.decode('utf-8'))
        self.assertEqual(response.code, 400)
        self.json_check(response)
```

…which fails and prints out:

```
STATUS:: 200
BODY:::: ""
```